llvm/lib/MC/MCC/AssemblerC.cpp

```diff
@@ -186,3 +193,118 @@ LLVMBool LLVMExceedsSizeLimitEraVM(LLVMMemoryBufferRef MemBuf,
+                                        uint64_t ReferencePC,
+                                        const char **ReferenceName) {
+  *ReferenceType = LLVMDisassembler_ReferenceType_InOut_None;
+  return nullptr;
+}
+
+LLVMBool LLVMDisassembleEraVM(LLVMTargetMachineRef T,
+                              LLVMMemoryBufferRef InBuffer, uint64_t PC,
+                              uint64_t Options, LLVMMemoryBufferRef *OutBuffer,
+                              char **ErrorMessage) {
+  TargetMachine *TM = unwrap(T);
+  const Triple &TheTriple = TM->getTargetTriple();
+  constexpr size_t InstrSize = 8;
+  constexpr size_t WordSize = 32;
+  constexpr size_t OutStringSize = 1024;
+  char OutString[OutStringSize];
+  MemoryBuffer *InMemBuf = unwrap(InBuffer);
+  const char *Bytes = InMemBuf->getBuffer().data();
+  size_t BytesNum = InMemBuf->getBufferSize();
+
+  if (PC > BytesNum) {
+    *ErrorMessage = strdup("Starting address exceeds the bytecode size");
+    return true;
+  }
+
+  if (PC % InstrSize) {
+    *ErrorMessage =
+        strdup("Starting address isn't multiple of 8 (instruction size)");
+    return true;
+  }
+
+  if (BytesNum % WordSize) {
+    *ErrorMessage = strdup("Bytecode size isn't multiple of 32 (word size)");
+    return true;
+  }
+
+  LLVMDisasmContextRef DCR = LLVMCreateDisasm(
+      TheTriple.getTriple().c_str(), nullptr, 0, nullptr, symbolLookupCallback);
+  assert(DCR && "Unable to create disassembler");
+
+  std::string Disassembly;
+  raw_string_ostream OS(Disassembly);
+  formatted_raw_ostream FOS(OS);
+  bool ShoulOutputEncoding =
+      Options & LLVMDisassemblerEraVM_Option_OutputEncoding;
+  auto PrintEncoding = [&FOS](uint64_t PC, uint8_t InstrBytes[InstrSize]) {
+    FOS << format("%8" PRIx64 ":", PC);
+    FOS << ' ';
+    dumpBytes(ArrayRef<uint8_t>(InstrBytes, InstrSize), FOS);
+  };
+
+  std::set<uint64_t> WordRefs;
+  SmallVector<uint8_t> CellBytes;
+  bool CollectCellBytes = false;
+  Regex CodeRegex("code\\[([0-9]+)\\]");
+  while (PC < BytesNum) {
+    uint8_t InstrBytes[InstrSize];
+    std::memcpy(InstrBytes, Bytes + PC, InstrSize);
+    size_t NumRead = LLVMDisasmInstruction(DCR, InstrBytes, InstrSize, /*PC=*/0,
+                                           OutString, OutStringSize);
+    // Check if the instruction contains a code reference. If so,
+    // extract the word number and add it to the WordRefs set.
+    SmallVector<StringRef, 2> Matches;
+    if (NumRead && CodeRegex.match(OutString, &Matches)) {
+      uint64_t WordNum = 0;
+      to_integer<uint64_t>(Matches[1], WordNum, /*Base=*/10);
+      WordRefs.insert(WordNum);
+    }
+    // If we are at the word boundary and the word is being referenced,
+    // emit the numeric label and the .cell directive. Also set up the
+    // flag that we are collecting the cell parts.
+    if (uint64_t WordNum = PC / WordSize;
+        (PC % WordSize) == 0 && WordRefs.count(WordNum)) {
+      CollectCellBytes = true;
```

Comment:
Can't we perform all data-related handling in a single `if` instead of introducing `CollectCellBytes`?

Reply:
Sure, will change this.